File: typescript/bob/bob.ts
function hasLetters(str: string): boolean {
  let regExp = /[a-zA-Z]/g;
  return regExp.test(str);
}

function isAllUpperCase(str: string): boolean {
  return str === str.toUpperCase();
}

function isSilent(str: string): boolean {
  return str.length === 0;
}

function isAskingQuestion(str: string): boolean {
  return str.slice(-1) === "?";
}

function isYelling(str: string): boolean {
  return hasLetters(str) && isAllUpperCase(str);
}

export function hey(message: string): string {
  message = message.trim();

  if (isSilent(message)) {
    return "Fine. Be that way!";
  } else if (isYelling(message) && isAskingQuestion(message)) {
    return "Calm down, I know what I'm doing!";
  } else if (isYelling(message)) {
    return "Whoa, chill out!";
  } else if (isAskingQuestion(message)) {
    return "Sure.";
  } else {
    return "Whatever.";
  }
}

File: ruby/bob/bob.rb
class Bob
  class << self
    def hey(remark)
      remark = Remark.new(remark)
      if remark.blank?
        "Fine. Be that way!"
      elsif remark.is_shouted_question?
        "Calm down, I know what I'm doing!"
      elsif remark.is_question?
        "Sure."
      elsif remark.is_shouted?
        "Whoa, chill out!"
      else
        "Whatever."
      end
    end
  end
end

class Remark
  attr_reader :text

  def initialize(text)
    @text = text.strip.squeeze(" ")
  end

  def blank?
    text.empty?
  end

  def is_shouted_question?
    is_question? && is_shouted?
  end

  def is_question?
    text[-1] == "?"
  end

  def is_shouted?
    text == text.upcase && contains_letters?
  end

  def contains_letters?
    text.match?(/[a-zA-Z]/)
  end
end
